src/vsphere_cpi/lib/cloud/vsphere/nsxt_provider.rb

```diff
@@ -118,42 +124,58 @@ def initialize(client, default_vif_type)
     def add_vm_to_nsgroups(vm, ns_groups)
       return if ns_groups.nil? || ns_groups.empty?
       return if nsxt_nics(vm).empty?
-
-      logger.info("Adding vm '#{vm.cid}' to NSGroups: #{ns_groups}")
-      nsgroups = retrieve_nsgroups(ns_groups)
-
       lports = logical_ports(vm)
-      nsgroups.each do |nsgroup|
-        logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
-        grouping_obj_svc.add_or_remove_ns_group_expression(
-          nsgroup.id,
-          *to_simple_expressions(lports),
-          'ADD_MEMBERS'
-        )
+      logger.info("Adding vm '#{vm.cid}' to NSGroups: #{ns_groups}")
+      Bosh::Retryable.new(
+          tries: 5,
+          on: [NSXTSecGrpUpdateError]
+      ).retryer do |i|
+        nsgroups = retrieve_nsgroups(ns_groups)
+        nsgroups.each do |nsgroup|
+          logger.info("Adding LogicalPorts: #{lports.map(&:id)} to NSGroup '#{nsgroup.id}'")
+          begin
+            grouping_obj_svc.add_or_remove_ns_group_expression(
+              nsgroup.id,
+              *to_simple_expressions(lports),
+              'ADD_MEMBERS'
+            )
+          rescue NSXT::ApiCallError => e
+            raise NSXTSecGrpUpdateError if e.code == 409 || e.code == 412 #Conflict or PreconditionFailed
```

Comment:
not sure if this will work. Shouldn't it be `raise NSXTSecGrpUpdateError.new("NSXT 409 db conflict or 412 revision conflict encountered with error :  #{e}")`

Reply:
Your code is raising a class without instantiating an object with `.new`. May be this works but I haven't seen before. Can you confirm?

Reply:
I looked into it and came across this - 
[https://stackoverflow.com/questions/9638475/raising-an-exception-use-instance-or-class](url)
I think this works fine. Do you think it would be cleaner to use an instance instead?

Reply:
In general in Ruby when raising an error `.new` shouldn't be used:

```ruby
raise NSXTSecGrpUpdateError, "..."
```

Raises an instance of `NSXTSecGrpUpdateError` with `message` set to "...". Here's the documentation for the `raise` method:

https://ruby-doc.org/core-2.7.0/Kernel.html#method-i-raise

Specifically the 2+ argument form of `raise` instantiates an instance of the error class if necessary and then sets the `mesg` instance variable on that instance to the second argument. This is what allows `Exception#to_s` to get the exception message:

https://github.com/ruby/ruby/blob/9e6e39c3512f7a962c44dc3729c98a0f8be90341/error.c#L1054

The crazy thing is that the `mesg` instance variable *is not* `@mesg`. This is one of those times in which the Ruby interpreter uses an instance variable name that *doesn't* begin with `@`. So there's actually no way for you to access it from Ruby land (you can get to it in C land).

Anyways this is a very long tangent to say that the preferred way in Ruby to raise an exception with a specific error message is with the 2+ argument form of `raise`.